File: sdk/FriendKit/Assets/FriendKit/Message.cs
using System.Collections;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Threading.Tasks;
using UnityEngine;

using Newtonsoft.Json;

namespace FriendKit
{
    public class MessageModel
    {
        public string id;

        public User sender;
        public User receiver;

        public string title;
        public string body;
    }
    public class Message : MessageModel
    {
        public async static Task<Message[]> Query(string name)
        {
            var func = FB.Functions;
            var resp = await func.GetHttpsCallable("message_query")
                .CallAsync(new Dictionary<string, object>()
                {
                });

            return resp.Data.Reinterpret<GetMessagesResponse>().messages;
        }

        public async Task Delete()
        {
            var func = FB.Functions;
            var resp = await func.GetHttpsCallable("message_delete")
                .CallAsync(new Dictionary<string, object>()
                {
                    ["mid"] = id
                });
        }
    }

    class GetMessagesResponse
    {
        public Message[] messages;
    }
}

File: functions/src/api/message.ts
import { onHttpsCall } from "../common";
import { NotAuthorizedError, PermissionDeniedError } from "../error";
import { User } from "../controller/user";
import { Message } from "../controller/message";

export const message_get = onHttpsCall(async (data, ctx) => {
    if (!ctx.auth) throw new NotAuthorizedError();

    const user = User.get(ctx.auth.uid);
    const messages = await Message.query(user);

    return {
        messages
    };
});
export const message_send = onHttpsCall(async (data, ctx) => {
    if (!ctx.auth) throw new NotAuthorizedError();

    const sender = User.get(ctx.auth.uid);
    const receiver = User.get(data.uid);
    const message = Message.create(sender, receiver, {
        body: data.body,
        title: data.title
    });

    return {
    };
});
export const message_delete = onHttpsCall(async (data, ctx) => {
    if (!ctx.auth) throw new NotAuthorizedError();

    const message = Message.get(data.mid);
    if (message.data!.receiverId !== ctx.auth.uid)
        throw new PermissionDeniedError();

    await message.delete();

    return {
    };
});